### ISSUE TITLE
Comparison targets are reversed (regression)

### ISSUE THREAD

**Issue by falsandtru** [Microsoft/TypeScript#31251]
The similar problem #30118 had been fixed, but appeared again.

**TypeScript Version:**  3.4.0-dev.20190504

**Search Terms:**

**Code**

https://github.com/falsandtru/spica/blob/v0.0.249/src/supervisor.legacy.ts#L56

**Expected behavior:**

pass

**Actual behavior:**

```
Argument of type 'this' is not assignable to parameter of type 'Supervisor<string, unknown, unknown, unknown>'.
  Type 'Supervisor<N, P, R, S>' is not assignable to type 'Supervisor<string, unknown, unknown, unknown>'.
    Type 'string' is not assignable to type 'N'.
      'string' is assignable to the constraint of type 'N', but 'N' could be instantiated with a different subtype of constraint 'string'.ts(2345)
```

**Playground Link:** 

**Related Issues:**

**Comment by falsandtru** [comment]
Note that a regression test for #30118 is already added but another case, posted here and there, is not covered by it.

**Comment by jack-williams** [comment]
Can't reproduce on `3.5.0-dev` - is this behaviour in a release branch or at some intermediate point?

**Comment by falsandtru** [comment]
Can you test with 3.5.0-dev.20190504?

**Comment by falsandtru** [comment]
ping @weswigham

**Comment by jack-williams** [comment]
Just tried off [master](https://github.com/Microsoft/TypeScript/commit/4ee0084fa1072b8217f303d17d301f9b9467fd69) and I **do** get the error you link.

I presume you are running `strictFunctionTypes`? 

This error looks correct to me. The property `instances` is a set of `Supervisor<string, unknown, unknown>`, therefore allowing the method `call` to accept arguments of type `string` in the `name` position. The `this` instance is generic in the `name` parameter to `call` and may be instantiated with something more specific like `"a"`, so it's not safe to assign `this` to `Supervisor<string, unknown, unknown>`.

A smaller example:
```ts
declare let withString: Supervisor<string>;
declare let withA: Supervisor<"a">;
withA.call("a", 'param'); // ok
withA.call("a string" as string, 'param'); // not ok
withString = withA; // correctly an error, otherwise we could make an illegal call through the alias
withString.call("a string" as string, 'param'); // no error
```

**Comment by jack-williams** [comment]
If you change `call` to be:
```ts
  public abstract call(name: ('' extends N ? undefined : never), param: P, timeout?: number): Promise<R>;
```
you still get the error, which is probably incorrect and due to an invariant variance calculation which has been turned on again for the conditional type extends type. Small repro:
```ts
interface A<T> {
    x: number extends T ? 1 : 1;
}

declare let a: A<number>;
declare let b: A<3>;

a = b; // error
b = a; // error
```

**Comment by falsandtru** [comment]
> I presume you are running strictFunctionTypes?

No: https://github.com/falsandtru/spica/blob/master/tsconfig.json

> declare let withString: Supervisor<string>;
declare let withA: Supervisor<"a">;
withA.call("a", 'param'); // ok
withA.call("a string" as string, 'param'); // not ok
withString = withA; // correctly an error, otherwise we could make an illegal call through the alias
withString.call("a string" as string, 'param'); // no error

TypeScript originally hasn't supported this problem as Arrays also have the same problem. Therefore type checking is wrong or wrongly too strict in this case.

>interface A<T> {
    x: number extends T ? 1 : 1;
}
>
>declare let a: A<number>;
declare let b: A<3>;
>
>a = b; // error
b = a; // error

Does it also fix this issue?

**Comment by jack-williams** [comment]
Yep a couple of things wrong in my comments. Corrections:
- `strictFunctionTypes` is irrelevant here because `call` is a method.
- Similarly, my example:
```ts
withString = withA; // correctly an error, otherwise we could make an illegal call through the alias
```
is not relevant because `call` is a method. It would be an error if `call` was an arrow function, but it's not and so should not error for your code.

> Does it also fix this issue?

The minor suggestion I posted does fix my small repro and your example - though I'm not sure if it's the correct thing to do overall.

**Comment by falsandtru** [comment]
@ahejlsberg Take a look?

**Comment by falsandtru** [comment]
@ahejlsberg Can you also fix this regression?
